### PR TITLE
Refactor requests throttling

### PR DESCRIPTION
**Pull request by Hirrolot**

### CODE REVIEW

File: src/adaptors/throttle.rs [hunk 1]
```diff
@@ -86,12 +86,14 @@ const DELAY: Duration = Duration::from_millis(250);
...
-    /// Allowed messages in one chat per minute
-    pub chat_m: u32,
+    /// Allowed messages in one chat per second.
+    pub messages_per_sec_chat: u32,
+
+    /// Allowed messages in one chat per minute.
+    pub messages_per_min_chat: u32,
+
+    /// Allowed messages per second.
+    pub messages_per_sec_overall: u32,
```
Comment (WaffleLapkin) [inline src/adaptors/throttle.rs:96]: What's the point in blank lines? Haven't seen blank lines in structures in our codebase before :thinking:

Comment (Hirrolot) [reply src/adaptors/throttle.rs:96]: With blank lines, the fields are easier to separate from each other. Compare with this:

```rust
/// Allowed messages in one chat per second.
pub messages_per_sec_chat: u32,
/// Allowed messages in one chat per minute.
pub messages_per_min_chat: u32,
/// Allowed messages per second.
pub messages_per_sec_overall: u32,
```

We already have blank lines in all the API types.

Comment (WaffleLapkin) [reply src/adaptors/throttle.rs:96]: Nvm then, I'm blind

**Comment by WaffleLapkin** [comment]
@Hirrolot to fix CI you can rebase:
```shell
$ git checkout master
$ git pull
$ git checkout refactor-throttle
$ git rebase master
$ git push origin refactor-throttle --force-with-lease 
```

File: src/adaptors/throttle.rs [hunk 2]
```diff
@@ -23,49 +23,48 @@ use crate::{
...
 //
-// 3. Record current time.
+// 3. Record the current time.
 //
-// 4. Clear history from records which time < (current - minute)
+// 4. Clear the history from records whose time < (current time - minute).
 //
-// 5. Count all requests in which were sent last second,
-//    `allowed = limit.overall_s - count`
+// 5. Count all requests which were sent last second, `allowed = limit.overall_s
```
Comment (WaffleLapkin) [inline src/adaptors/throttle.rs:55]:
```suggestion
// 5. Count all requests which were sent last second, `allowed = limit.messages_per_sec_overall
```

File: src/adaptors/throttle.rs [hunk 3]
```diff
@@ -149,48 +150,34 @@ impl Default for Limits {
...
-            match queue_rx.recv().now_or_never() {
-                Some(Some(req)) => queue.push(req),
-                // There are no items in queue
-                None => break,
-                // The queue was closed
-                Some(None) => close = true,
-            }
-        }
+    let mut when_requests_were_sent: VecDeque<(FastChatId, Instant)> = VecDeque::new();
+    let mut requests_sent_to_chats = RequestsSentToChats::default();
```
Comment (WaffleLapkin) [inline src/adaptors/throttle.rs:175]: IMO, this naming is **too** verbose. I think identifiers this long are not helping to read code but instead are overwhelming the reader.

I would better use `history` and `sent_requests`. Some points in addition to the length:
- `when_requests_were_sent` doesn't suggest that the deque is sorted by time
- all requests are sent to chats in some sense...

File: src/adaptors/throttle.rs [hunk 4]
```diff
@@ -251,75 +241,93 @@ async fn worker(limits: Limits, mut queue_rx: mpsc::Receiver<(Id, Sender<Never>)
...
-                    drop(entry.remove());
+            *requests_sent_to_chats.per_sec.entry(*chat).or_insert(0) += 1;
+        }
 
-                    // We've "sent" 1 request, so now we can send 1 less
-                    allowed -= 1;
-                    if allowed == 0 {
-                        break;
-                    }
+        let mut queue_removing_iter = queue.removing();
```
Comment (WaffleLapkin) [inline src/adaptors/throttle.rs:263]: It's not really an iter though :thinking:

File: src/adaptors/throttle.rs [hunk 5]
```diff
@@ -251,75 +241,93 @@ async fn worker(limits: Limits, mut queue_rx: mpsc::Receiver<(Id, Sender<Never>)
...
+
+        while let Some(entry) = queue_removing_iter.next() {
+            let chat = &entry.value().0;
+            let messages_sent = requests_sent_to_chats
```
Comment (WaffleLapkin) [inline src/adaptors/throttle.rs:267]: `requests_sent` is a more appropriate name.

File: src/adaptors/throttle.rs [hunk 6]
```diff
@@ -484,34 +492,34 @@ download_forward! {
...
 
-/// Id used in worker.
+/// An ID used in the worker.
 ///
 /// It is used instead of `ChatId` to make copying cheap even in case of
-/// usernames. (It just hashes username)
+/// usernames. (It is just a hashed username.)
 #[derive(Debug, Copy, Clone, Hash, Eq, PartialEq)]
-enum Id {
+enum FastChatId {
```
Comment (WaffleLapkin) [inline src/adaptors/throttle.rs:500]: I think `ChatIdHash` would better describe this type.

**Review by WaffleLapkin — CHANGES_REQUESTED** [review]

File: src/adaptors/throttle.rs (hunk 4 above)
Comment (Hirrolot) [reply src/adaptors/throttle.rs:263]: Why this is not an iterator? We call `.next()` then.

Comment (WaffleLapkin) [reply src/adaptors/throttle.rs:263]: because it doesn't implement `Iterator`

**Review by WaffleLapkin — APPROVED** [review]

**Pull request merged**